scripts/sign_rproc_fw.py

```diff
@@ -0,0 +1,568 @@
+#!/usr/bin/env python3
+# SPDX-License-Identifier: BSD-2-Clause
+#
+# Copyright (C) 2023, STMicroelectronics - All Rights Reserved
+#
+
+try:
+    from elftools.elf.elffile import ELFFile
+    from elftools.elf.sections import SymbolTableSection
+    from elftools.elf.enums import ENUM_P_TYPE_ARM
+    from elftools.elf.enums import *
+except ImportError:
+    print("""
+***
+ERROR: "pyelftools" python module is not installed or version < 0.25.
+***
+""")
+    raise
+
+try:
+    from Cryptodome.Hash import SHA256
+    from Cryptodome.Signature import pkcs1_15
+    from Cryptodome.PublicKey import RSA
+    from Cryptodome.Signature import DSS
+    from Cryptodome.PublicKey import ECC
+except ImportError:
+    print("""
+***
+ERROR: "pycryptodomex" python module should be installed.
+***
+""")
+    raise
+
+import os
+import sys
+import struct
+import logging
+import binascii
+
+#  Generated file structure:
+#
+#                    ----+-------------+
+#                   /    |    Magic    |
+#                  /     +-------------+
+#                 /      +-------------+
+#                /       |   version   |
+#               /        +-------------+
+#              /         +-------------+
+# +-----------+          | sign size   |   size of the signature
+# |   Header  |          +-------------+   (in bytes, 64-bit aligned)
+# +-----------+          +-------------+
+#              \         | images size |   size of the images to load
+#               \        +-------------+   (in bytes, 64-bit aligned)
+#                \       +-------------+
+#                 \      |  TLV size   |   Generic TLV size
+#                  \     +-------------+   (in bytes, 64-bit aligned)
+#                   \    +-------------+
+#                    \   | PLAT TLV sz |   Platform TLV size
+#                     \--+-------------+   (in bytes, 64-bit aligned)
+#                        +-------------+
+#                        |     TLV     |   Information used to authenticate the
+#                        |             |   firmware, stored in
+#                        +-------------+   Type-Length-Value format.
+#                        +-------------+
+#                        |   Platform  |   Specific platform information,
```

Comment:
Why do the Platform TLVs need to be separated from the other TLVs? By reserving a type namespace for each they could be kept apart in that way instead when needed.

Reply:
As I see it you're already using two namespaces `PLAT_TLV_PREDEF_TYPES` uses a `0x0000002` prefix and `TLV_TYPES` anything below `0x00000020`. I suppose that could be improved by giving larger ranges, but the principle is there already. By looking at a type field you can already tell Platform TLVs apart from the others.

Reply:
The `PLAT_TLV_PREDEF_TYPES `and `PLAT_TLV_TYPES` are 2 independent structure.

- The `PLAT_TLV_TYPES` is used by the `TLV` class
- The PLAT_TLV_PREDEF_VALUES is used by the `custom_tlv` class. It lists predefined TLV for platforms. The offset x 0x20 (range could be higner) is use to give a namespace starting at 0x0 for platform TLVs. User can add platform TLV in argument of the script, either using prredefined type or giving it own type ID defined in the platform PTA.

To be honest, I wonder if having predefined type for platform TLV in the TA makes sense...

Reply:
I'd rather drop the "platform TLV" area and only have a single TLV area. To make it extendable you could reserve the 0x0000_FFxx range of types for extensions, while the 0x0000_00xx range is for official use and the rest are reserved for future use. Upstream OP-TEE should typically only use the official range of types as a way of keeping fragmentation in check.

Reply:
TLV type from 0x00020000 to 0xFFFF_FFFF should be reserved.

The TLV types sent to the PTA should probably also be official and standardized in some way.
Perhaps we need one more range for unofficial TLV types for downstream use, but this just a small detail that we can work out later if needed.

Reply:
I confirm that we will need flexibility to add unofficial TLV for downstream as we address Mass Market. We are already considering using this platform TLV to provide some boot configurations.
In short term mandatory platform TLVs are the boot address of the secure and non secure Firmware.

Standardize platform TLV would means list all of them, right? As they are platform dependent that would means having an exhaustive list of platform TLV taking account all platforms needs/constraints.  In this case I assume that  both the script and the remoteproc core TA should know this list.

An alternative would have only range for platform TLV and each platform define is own offical TLV in its PTA in upstream.
In downstream if extra plat TLV is added,  we can assume that the platform PTA will be impacted to add the new TLV definition in the list but also associated functions to treat it. 
The script today is oriented to this alternative using the " "--plat-tlv" parameter that add a platform TLV in the list without checking the match between the type and the associate data. the control will be done in